OpenSim/Simulation/StatesTrajectory.h

```diff
@@ -0,0 +1,641 @@
+ * model.initSystem();
+ * double knee_angle = model.getStateVariableValue(states[0], "knee/flexion/value");
+ * Vec3 com = model.calcMassCenterPosition(states[0]);
+ * @endcode
+ *
+ * Depending on the quantity you want to obtain, you may also need to realize
+ * the state to a certain stage:
+ * @code{.cpp}
+ * model.getMultibodySystem().realize(states[0], SimTK::Stage::Velocity);
+ * model.getMuscles().get("soleus_r").getActivation(states[0]);
+ * @endcode
+ *
+ * You can iterate through a trajectory to access the value of a state variable
+ * at each time in the trajectory.
+ * @code{.cpp}
+ * for (const auto& state : states) {
+ *     std::cout << state.getTime() << " "
+ *               << model.getStateVariableValue(state, "knee/flexion/value")
+ *               << std::endl;
+ * }
+ * @endcode
+ *
+ */
+class OSIMSIMULATION_API StatesTrajectory {
+public:
+    /** Create an empty trajectory of states. */
+    StatesTrajectory() {}
+
+    /** The number of SimTK::State%s in the trajectory. */
+    size_t getSize() const;
+
+    /// @name Accessing individual SimTK::State%s
+    /// @{
+    /** Get a const reference to the state at a given index in the trajectory.
+     * Here's an example of getting a state variable value from the first state
+     * in the trajectory: 
+     * @code{.cpp}
+     * Model model("subject01.osim");
+     * const StatesTrajectory states = getStatesTrajectorySomehow();
+     * const auto& state = states[0];
+     * model.getStateVariableValue(state, "knee/flexion/value");
+     * @endcode
+     * This function does not check if the index is larger than the size of
+     * the trajectory; see get() if you want this check. */
+    const SimTK::State& operator[](size_t index) const {
+        return m_states[index];
+    }
+    /** Get a const reference to the state at a given index in the trajectory.
+
+     * @throws IndexOutOfRange If the index is greater than the size of the
+     *                         trajectory.
+     */
+    // TODO throw OpenSIm IndexOutOfRange exception, and update Doxygen to
+    // remove out_of_range.
+    const SimTK::State& get(size_t index) const {
+        try {
+            return m_states.at(index);
+        } catch (const std::out_of_range&) {
+            OPENSIM_THROW(IndexOutOfRange, index, 0, m_states.size() - 1);
+        }
+    }
+    /** Get a const reference to the first state in the trajectory. */
+    const SimTK::State& front() const { 
+        return m_states.front();
+    }
+    /** Get a const reference to the last state in the trajectory. */
+    const SimTK::State& back() const { 
+        return m_states.back();
+    }
+
+    /** Get the index of the state just before (or at) the given time.
+     * @throws TimeOutOfRange If there are no states before the given time.
+     * @throws Exception If the trajectory is empty.
+     * @see getIndexBefore(const double&, const double&)
+     */
+    size_t getIndexBefore(const double& time) {
```

Comment:
The use of separate methods for the tolerance was intentional. I think it's rare that someone wants to use the tolerance, and thus that in most cases it would only make the methods look more complicated than they are. If the users run into some unexpected behavior, they can look at the doxygen again and find out that there's another variant (that accepts a tolerance) to help them. This is distinct from the `createFromStatesStorage()` method, where I think the default arguments are fairly easy to comprehend and that users are likely to want to alter them.